### PR TITLE
Refactoring of events for variables creation, update and delete

### PR DESCRIPTION
**Pull request by CTI777**
Related to https://github.com/Activiti/Activiti/issues/2724
Depends on 
https://github.com/Activiti/activiti-api/pull/119
https://github.com/Activiti/activiti-api/pull/120

### CODE REVIEW

**Comment by CTI777** [comment]
@salaboy @erdemedeiros 
After checking handling of task variable deletion following interesting things are found to discuss:

- **If we delete a task variable we set a flag that it is deleted**  - do we really need to throw an exception if variable is already deleted 

- **If we create a variable there is no check that variable already exists**: have added a check here

- **If we update a variable we do not check if it is already deleted**: have added the check here

- **When we query variables, all variables are returned (deleted and existing)**: will it have a sense to return always (or by default) only existing variables (do not include deleted variables at all). Otherwise it is possible that a value of the deleted variable will be used…

**Comment by CTI777** [comment]
@salaboy @erdemedeiros @igdianov 
Handling events related to task variables (as an example) may lead sometime to exceptions (not found variable, variable already exists, etc.) Catching of such kind of exceptions is not implemented, and this leads for repetitive sending of events with following exceptions again.. Even testing becomes painful when some messages are hanging in RabbitMQ and suddenly appears in unpredictable places. 
**Can we once discuss such kind of problems**?
Thrown exceptions (which are correct from business logic) will be accumulated in production system and may slow down everything dramatically...

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/ProcessVariableUpdateEventHandler.java [hunk 1]
```diff
@@ -34,6 +34,9 @@ public void handle(ProcessVariableEntity updatedVariableEntity) {
         BooleanExpression predicate = QProcessVariableEntity.processVariableEntity.name.eq(variableName)
                 .and(
                         QProcessVariableEntity.processVariableEntity.processInstanceId.eq(String.valueOf(processInstanceId))
+                )
+                .and(
+                        QProcessVariableEntity.processVariableEntity.markedAsDeleted.eq(Boolean.FALSE)
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/ProcessVariableUpdateEventHandler.java]: As discussed previously, let's stop using `markedAsDeleted` flag and use real deletes.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskVariableDeletedEventHandler.java [hunk 2]
```diff
@@ -43,9 +53,27 @@ public void handle(CloudVariableDeletedEvent event) {
...
-                                                             "Unable to find variableEntity with name '" + variableName + "' for task '" + taskId + "'");
+                                                                 predicate,
+                                                                 "Unable to find variableEntity with name '" + variableName + "' for task '" + taskId + "'");
+        
+        predicate = QTaskEntity.taskEntity.id.eq(taskId);
+        
+        Optional<TaskEntity> taskEntity = taskRepository.findOne(predicate);
+            
+        if (taskEntity.isPresent()) {
+            TaskStatus taskStatus = taskEntity.get().getStatus();
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskVariableDeletedEventHandler.java]: Reuse `taskEntity.get().getStatus().isFinalStatus()`. See https://github.com/Activiti/activiti-cloud-query-service/pull/231#discussion_r284597687.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskVariableUpdatedEventHandler.java [hunk 3]
```diff
@@ -31,9 +31,12 @@ public TaskVariableUpdatedEventHandler(TaskVariableUpdater variableUpdater) {
     public void handle(TaskVariableEntity updatedVariableEntity) {
         String variableName = updatedVariableEntity.getName();
         String taskId = updatedVariableEntity.getTaskId();
-        BooleanExpression predicate = QTaskVariableEntity.taskVariableEntity.name.eq(variableName)
+        BooleanExpression predicate = QTaskVariableEntity.taskVariableEntity.name.eq(String.valueOf(variableName))
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskVariableUpdatedEventHandler.java]: Are you sure this extra `String.valueOf` is necessary here? `variableName` is already a String.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskVariableUpdatedEventHandler.java [hunk 4]
```diff
@@ -31,9 +31,12 @@ public TaskVariableUpdatedEventHandler(TaskVariableUpdater variableUpdater) {
...
                 .and(
                         QTaskVariableEntity.taskVariableEntity.taskId.eq(String.valueOf(taskId))
+                )
+                .and(
+                        QTaskVariableEntity.taskVariableEntity.markedAsDeleted.eq(Boolean.FALSE)
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/TaskVariableUpdatedEventHandler.java]: No longer necessary.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/VariableCreatedEventHandler.java [hunk 5]
```diff
@@ -69,19 +73,35 @@ public void handle(CloudRuntimeEvent<?, ?> event) {
...
         TaskEntity taskEntity = entityManager.getReference(TaskEntity.class,
-                                                           variableCreatedEvent.getEntity().getTaskId());
+                                                           taskId);
+           
+        BooleanExpression predicate = QTaskVariableEntity.taskVariableEntity.name.eq(variableName)
+                .and(
+                        QTaskVariableEntity.taskVariableEntity.taskId.eq(String.valueOf(taskId))
+                )
+                .and(
+                        QTaskVariableEntity.taskVariableEntity.markedAsDeleted.eq(Boolean.FALSE)
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/VariableCreatedEventHandler.java]: No longer necessary.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/VariableCreatedEventHandler.java [hunk 6]
```diff
@@ -69,19 +73,35 @@ public void handle(CloudRuntimeEvent<?, ?> event) {
...
+                );
+
+        //Should we throw exception here?
+        if (taskVariableRepository.exists(predicate)) {
+            throw new IllegalStateException("Variable " + variableName + " already exists in the task " + taskId + "!");
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/VariableCreatedEventHandler.java]: I think in this case you can log a message and ignore the event.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/VariableCreatedEventHandler.java [hunk 7]
```diff
@@ -92,16 +112,29 @@ private void createTaskVariableEntity(CloudVariableCreatedEvent variableCreatedE
...
         ProcessInstanceEntity processInstanceEntity= getProcessInstance(variableCreatedEvent);
+        String processInstanceId = variableCreatedEvent.getEntity().getProcessInstanceId();
+        String variableName = variableCreatedEvent.getEntity().getName();
+        
+        BooleanExpression predicate = QProcessVariableEntity.processVariableEntity.name.eq(variableName)
+                .and(
+                        QProcessVariableEntity.processVariableEntity.processInstanceId.eq(String.valueOf(processInstanceId))
+                )
+                .and(
+                        QProcessVariableEntity.processVariableEntity.markedAsDeleted.eq(Boolean.FALSE)
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/VariableCreatedEventHandler.java]: No longer necessary.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/VariableCreatedEventHandler.java [hunk 8]
```diff
@@ -92,16 +112,29 @@ private void createTaskVariableEntity(CloudVariableCreatedEvent variableCreatedE
...
+                );
         
+        if (variableRepository.exists(predicate)) {
+            throw new IllegalStateException("Variable " + variableName + " already exists in the process " + processInstanceId + "!");
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/VariableCreatedEventHandler.java]: I think you can log the exception and ignore the event in this case.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/ProcessInstanceVariableAdminController.java [hunk 9]
```diff
@@ -66,15 +66,19 @@ public ProcessInstanceVariableAdminController(VariableRepository variableReposit
                                                                         Pageable pageable) {
 
         QProcessVariableEntity variable = QProcessVariableEntity.processVariableEntity;
-        BooleanExpression expression = variable.processInstanceId.eq(processInstanceId);
         
-        Predicate extendedPredicated = expression;
+        //We will show only not deleted variables 
+        BooleanExpression expression = variable.processInstanceId.eq(processInstanceId)
+                                       .and(variable.markedAsDeleted.eq(Boolean.FALSE));
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/ProcessInstanceVariableAdminController.java]: No longer necessary.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/ProcessInstanceVariableController.java [hunk 10]
```diff
@@ -67,12 +67,17 @@ public ProcessInstanceVariableController(ProcessInstanceVariableResourceAssemble
                                                                         Pageable pageable) {
 
         QProcessVariableEntity variable = QProcessVariableEntity.processVariableEntity;
-        BooleanExpression expression = variable.processInstanceId.eq(processInstanceId);
+        
+        //We will show only not deleted variables 
+        BooleanExpression expression = variable.processInstanceId.eq(processInstanceId)
+                                       .and(variable.markedAsDeleted.eq(Boolean.FALSE));
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/ProcessInstanceVariableController.java]: No longer necessary.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskVariableAdminController.java [hunk 11]
```diff
@@ -67,14 +67,17 @@ public TaskVariableAdminController(TaskVariableRepository variableRepository,
                                                                         Pageable pageable) {
 
         QTaskVariableEntity variable = QTaskVariableEntity.taskVariableEntity;
-        BooleanExpression expression = variable.taskId.eq(taskId);
 
-        Predicate extendedPredicated = expression;
+        //We will show only not deleted variables 
+        BooleanExpression expression = variable.taskId.eq(taskId)
+                                       .and(variable.markedAsDeleted.eq(Boolean.FALSE));
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskVariableAdminController.java]: No longer necessary.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskVariableController.java [hunk 12]
```diff
@@ -68,12 +68,17 @@ public TaskVariableController(TaskVariableRepository variableRepository,
                                                                         Pageable pageable) {
 
         QTaskVariableEntity variable = QTaskVariableEntity.taskVariableEntity;
-        BooleanExpression expression = variable.taskId.eq(taskId);
+        
+        //We will show only not deleted variables 
+        BooleanExpression expression = variable.taskId.eq(taskId)
+                                       .and(variable.markedAsDeleted.eq(Boolean.FALSE));
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskVariableController.java]: No longer necessary.

File: activiti-cloud-services-query/activiti-cloud-services-query-rest/src/test/java/org/activiti/cloud/services/query/events/handlers/TaskEntityVariableEntityDeletedEventHandlerTest.java [hunk 13]
```diff
@@ -66,7 +72,11 @@ public void handleShouldSoftDeleteIt() {
         TaskVariableEntity variableEntity = new TaskVariableEntity();
         given(entityFinder.findOne(eq(variableRepository), any(Predicate.class), anyString())).willReturn(variableEntity);
 
-
+        TaskEntity taskEntity = new TaskEntity();
```
Comment (erdemedeiros) [inline activiti-cloud-services-query/activiti-cloud-services-query-rest/src/test/java/org/activiti/cloud/services/query/events/handlers/TaskEntityVariableEntityDeletedEventHandlerTest.java:74]: This test should be renamed as we are no longer using soft deletes.

**Review by erdemedeiros — CHANGES_REQUESTED** [review]

**Comment by CTI777** [comment]
@erdemedeiros @salaboy 
Have made refactoring. 
For task variables status of the task is checked (and nothing is deleted if task ended)
For process variables status of the process is not checked, variables are always deleted
Is this Ok?
If not Ok, 
maybe we can do following:
If process is completed: mark process variables as Deleted (soft deletion)

**Review by salaboy — APPROVED** [review]

**Pull request merged**